### ISSUE TITLE
Implementing draggable points

### ISSUE THREAD

**Issue by davecranwell** [alex3165/react-mapbox-gl#212]
I'm on a vanilla JS project which i'm porting to React, using react-mapbox-gl, where some markers in a layer need to be dragged with the mouse to reposition them.

My vanilla project implemented something similar to Mapbox's example here https://www.mapbox.com/mapbox-gl-js/example/drag-a-point/, but I'm wondering how you suggest I go about doing this in React. Perhaps you've already had some thoughts. Ideally I'd like to contribute something with a PR once I'm happy with it.

My idea was that a `<Layer>` might contain a `<Draggable>` that could wrap the children `<Feature>s`.

One problem I'd experienced trying this approach was that Layer IDs didn't seem to be available to children. So the `<Draggable>` couldn't use `map.queryRenderedFeatures(e.lngLat, { layers: ['layerId'] });` to limit results to just the markers I needed.

Any help appreciated.

**Comment by alex3165** [comment]
Hey @davecranwell sorry for the late reply, that's a very interesting feature, I would love to see something like this on `react-mapbox-gl`. Having a `Draggable` component that you use as a children of the `Layout` seems a good approach to me. you can pass down the layerId by cloning the children component of the layer and injecting the `layerId` as a prop http://reactjs.cn/react/docs/top-level-api.html#react.cloneelement.

**Comment by alex3165** [comment]
Would be happy to help you on this one as I think it is a very valuable feature to have !

**Comment by alex3165** [comment]
Feature is now draggable with version `2.0.3`, you can try out with the `switch-style` example

**Comment by davecranwell** [comment]
Awesome, thanks! I implemented it much the same as you. 

One thing I've discovered is that on full-screen maps or those with lots of other points in the layer, mapbox struggles to keep 60fps as you drag the marker around, causing it to lag considerably behind the pointer. I suspect mapbox's engine just isn't optimised for that kind of speed of input.

Overlays, however - or in this case Markers - would perform far better I suspect. I'm going to try applying a drag feature to those next.
